client/fed_linux_test.go

```diff
@@ -580,9 +579,67 @@ func TestSyncUpload(t *testing.T) {
 		verifySuccessfulTransfer(t, transferDetailsDownload)
 
 		// Verify we received the original contents, not any modified contents
-		contentBytes, err := os.ReadFile(filepath.Join(downloadDir, dirName, filepath.Base(innerTempDir), filepath.Base(innerTempFile.Name())))
+		contentBytes, err := os.ReadFile(filepath.Join(downloadDir, filepath.Base(innerTempDir), filepath.Base(innerTempFile.Name())))
 		require.NoError(t, err)
-		assert.Equal(t, innerTestFileContent, string(contentBytes))
+		require.Equal(t, innerTestFileContent, string(contentBytes))
+	})
+
+	t.Run("testSyncUploadFile", func(t *testing.T) {
+		// Create a new test file to upload
+		newDir := t.TempDir()
+		newTestFile, err := os.CreateTemp(newDir, "newTest")
+		newTestFileContent := "This is a brand new file"
+		require.NoError(t, err, "Error creating new test file")
+		_, err = newTestFile.WriteString(newTestFileContent)
+		require.NoError(t, err, "Error writing to new test file")
+
+		dirName := filepath.Base(tempDir)
+		uploadUrl := fmt.Sprintf("pelican://%s/first/namespace/sync_upload_none/%s/%s", discoveryUrl.Host, dirName, "test_single_upload")
+
+		// Upload the file
+		transferDetailsUpload, err := client.DoPut(fed.Ctx, newTestFile.Name(), uploadUrl, true, client.WithTokenLocation(tempToken.Name()), client.WithSynchronize(client.SyncSize))
+		require.NoError(t, err)
+		require.Len(t, transferDetailsUpload, 1)
+
+		// Download the new object
+		downloadDir := t.TempDir()
+		transferDetailsDownload, err := client.DoGet(fed.Ctx, uploadUrl, downloadDir, true, client.WithTokenLocation(tempToken.Name()), client.WithSynchronize(client.SyncSize))
+		require.NoError(t, err)
+		require.Len(t, transferDetailsDownload, 1)
+
+		// Verify we received the new contents with the expected object name
+		contentBytes, err := os.ReadFile(filepath.Join(downloadDir, "test_single_upload"))
+		require.NoError(t, err)
+		require.Equal(t, newTestFileContent, string(contentBytes))
+
+		smallTestFile, err := os.CreateTemp(newDir, "smallTest")
+		smallTestFileContent := "smaller test file"
+		require.NoError(t, err, "Error creating small test file")
+		_, err = smallTestFile.WriteString(smallTestFileContent)
+		require.NoError(t, err, "Error writing to small test file")
+
+		//Upload the file into the same location as the previous test file - should overwrite
```

Comment:
Doesn't overwriting an object violate its immutability? Shouldn't that be something we disallow?

Reply:
I'll talk with Brian in this case, but in our original discussion and my pictures of the whiteboard, he wanted an overwrite because the idea is that we function mostly like `rsync`. 

I think the case isn't that we don't allow overwrites, we just assume immutability and download behavior is undefined if the origin value is changed. 

Since the origin owner/admin could easily change the underlying files on a POSIX origin as well, I think preventing it in this case isn't worth it unless we could also protect the underlying file structure, which we can't.

Reply:
Actually, even more importantly, if the sync fails midway, we need to be able to overwrite the file even if it still exists. Because the original failed. So yes, we need to be able to do this.